README
NAME
    Script::Ichigeki - Perl extention for one time script.

VERSION
    This document describes Script::Ichigeki version 0.01.

SYNOPSIS
        use Script::Ichigeki;

    It is same as

        use Script::Ichigeki (
            exec_date       => 'XXXX-XX-XX', # today
            confirm_dialog  => 1,
            dialog_message  => 'Do you really execute `%s` ?',
        );

    or

        use Script::Ichigeki ();
        Script::Ichigeki->hissatsu(
            exec_date       => 'XXXX-XX-XX', # today
            confirm_dialog  => 1,
            dialog_message  => 'Do you really execute `%s` ?',
        );

DESCRIPTION
    Script::Ichigeki is the module for one time script for mission critical
    (especially for preventing rerunning it).

    Only describing `use Script::Ichigeki`, confirm dialog is displayed and
    execution result is saved in log file automatically. This log file
    double with lock file for mutual exclusion and preventing rerunning.

CAUTION
    THE SOFTWARE IS IT'S IN ALPHA QUALITY. IT MAY CHANGE THE API WITHOUT
    NOTICE.

    If forking in your script, the software may not be properly handling it.

INTERFACE
  Functions
   "hissatsu(%options)"
    Automatically called in use phase.

    Available options are:

   "exec_date => 'Str|Time::Piece'"
    Date for execution date. Format is '%Y-%m-%d'.

   "confirm_dialog => 'Bool'"
    Confirm dialog is to be displayed or not. default: 1

   "dialog_message => 'Str'"
    Message of confirm dialog. Script name is expanded to '%s'. If using
    multibyte strings, you should "$ use utf8;" before "$ use
    Script::Ichigeki". default: 'Do you really execute `%s` ?',

DEPENDENCIES
    Perl 5.8.1 or later.

BUGS
    All complex software has bugs lurking in it, and this module is no
    exception. If you find a bug please either email me, or add the bug to
    cpan-RT.

SEE ALSO
    perl

AUTHOR
    Masayuki Matsuki <y.songmu@gmail.com>

LICENSE AND COPYRIGHT
    Copyright (c) 2012, Masayuki Matsuki. All rights reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Script::Ichigeki

0.03    Fri Dec 14 02:18:00 2012
        - add dialog_message feature
        - fix some broken English

0.02    Thu Dec 13 13:13:00 2012
        - fix awful English.

0.01    Thu Dec 13 11:52:00 2012
        - original version


Build.PL
use strict;
use warnings;
use Module::Build;

my $build = Module::Build->new(
    license        => 'perl',
    dynamic_config => 0,
    dist_abstract  => 'Perl extention for one time script.',

    build_requires     => {
        'Test::More'     => '0.98',
        'Test::Requires' => '0.06',
    },
    configure_requires => { 'Module::Build' => '0.38' },
    requires           => {
        perl         => '5.008001',
        'Mouse'         => 0,
        'Time::Piece'   => 0,
        'Path::Class'   => 0,
        'File::Tee'             => 0,
        'IO::Prompt::Simple'    => 0,
        'Term::Encoding'        => 0,
    },

    no_index    => { 'directory' => [ 'inc' ] },
    name        => 'Script/Ichigeki',
    module_name => 'Script::Ichigeki',

    # script_files => [''],

    test_files           => (-d '.git' || $ENV{RELEASE_TESTING}) ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    create_makefile_pl => 'traditional',
    create_readme      => 1,
    create_license     => 1,
);
$build->create_build_script;


lib/Script/Ichigeki/Hissatsu.pm
package Script::Ichigeki::Hissatsu;
use Mouse;
use Mouse::Util::TypeConstraints;

use Time::Piece;
use Path::Class qw/file/;
use IO::Prompt::Simple qw/prompt/;
use IO::Handle;
use File::Tee qw/tee/;

subtype 'Time::Piece' => as Object => where { $_->isa('Time::Piece') };
coerce 'Time::Piece'
    => from 'Str',
    => via {
        my $t = Time::Piece->strptime($_, '%Y-%m-%d');
        die "Invalie time format: [$_] .(format should be '%Y-%m-%d'.)" unless $t;
        localtime($t);
    };

has exec_date => (
    is      => 'ro',
    isa     => 'Time::Piece',
    coerce => 1,
    default => sub {
        localtime(Time::Piece->strptime(localtime->ymd, "%Y-%m-%d"));
    }
);

has confirm_dialog => (
    is      => 'ro',
    default => 1,
);

has log_file_postfix => (
    is      => 'ro',
    default => '.log',
);

has script => (
    is       => 'ro',
    default  => sub { file($0) },
);

has is_running => (
    is       => 'rw',
);

has in_compilation => (
    is => 'ro'
);

no Mouse;

sub execute {
    my $self = shift;

    my $now   = localtime;
    my $today = localtime(Time::Piece->strptime($now->ymd, "%Y-%m-%d"));
    $self->_exiting('exec_date: '. $self->exec_date->strftime('%Y-%m-%d') .' is not today!') unless $self->exec_date == $today;

    $self->_exiting(sprintf('execute log file [%s] alredy exists!', $self->_log_file)) if -f $self->_log_file;

    if ($self->confirm_dialog) {
        my $answer = prompt('Do you really execute `' . $self->script->basename . '` ? (y/n)');
        $self->_exiting('canceled.') unless $answer =~ /^y(?:es)?$/i;
    }

    STDOUT->autoflush;
    STDERR->autoflush;

    $self->_log(join "\n",
        '# This file is generated dy Script::Icigeki.',
        "start: @{[localtime->datetime]}",
        '---', ''
    );

    $self->is_running(1);
    tee STDOUT, $self->_log_fh;
    tee STDERR, $self->_log_fh;
}

{
    my  $_log_file;
    sub _log_file {
        my $self = shift;
        $_log_file ||= do {
            my $script = $self->script;
            $script->dir->file('.' . $script->basename . $self->log_file_postfix);
        };
    }

    my $_log_fh;
    sub _log_fh {
        $_log_fh ||= shift->_log_file->open('>>');
    }
}

sub _log {
    shift->_log_fh->print(@_);
}


sub _exiting {
    my ($self, $msg) = @_;

    $msg .= "\n";
    if ($self->in_compilation) {
        warn $msg;
        exit 1;
    }
    else {
        die $msg;
    }
}

sub DEMOLISH {
    my $self = shift;
    if ($self->is_running) {
        my $now = localtime->datetime;
        $self->_log(join "\n",
            '','---',
            "end: $now",'',
        );
    }
}

__PACKAGE__->meta->make_immutable;

1;
__END__

=head1 NAME

Script::Ichigeki::Hissatsu - Perl extention to do something

=head1 VERSION

This document describes Script::Ichigeki::Hissatsu version 0.01.

=head1 SYNOPSIS

    use Script::Ichigeki::Hissatsu;

=head1 DESCRIPTION

# TODO

=head1 INTERFACE

=head2 Functions

=head3 C<< execute(%options) >>

=head1 DEPENDENCIES

Perl 5.8.1 or later.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

Masayuki Matsuki E<lt>y.songmu@gmail.comE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2012, Masayuki Matsuki. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Script/Ichigeki.pm
package Script::Ichigeki;
use 5.008_001;
use strict;
use warnings;
our $VERSION = '0.03';

use Script::Ichigeki::Hissatsu;

sub import {
    my $pkg = shift;
    $pkg->hissatsu(@_, in_compilation => 1);
}

{
    my $_HISSATSU;
    sub hissatsu {
        die "Already running!\n" if $_HISSATSU;
        my $pkg = shift;

        $_HISSATSU = Script::Ichigeki::Hissatsu->new(@_);
        $_HISSATSU->execute;
    }

    END {
        undef $_HISSATSU;
    }
}

1;
__END__

=head1 NAME

Script::Ichigeki - Perl extention for one time script.

=head1 VERSION

This document describes Script::Ichigeki version 0.01.

=head1 SYNOPSIS

    use Script::Ichigeki;

It is same as

    use Script::Ichigeki (
        exec_date       => 'XXXX-XX-XX', # today
        confirm_dialog  => 1,
        dialog_message  => 'Do you really execute `%s` ?',
    );

or

    use Script::Ichigeki ();
    Script::Ichigeki->hissatsu(
        exec_date       => 'XXXX-XX-XX', # today
        confirm_dialog  => 1,
        dialog_message  => 'Do you really execute `%s` ?',
    );

=head1 DESCRIPTION

Script::Ichigeki is the module for one time script for mission critical
(especially for preventing rerunning it).

Only describing `use Script::Ichigeki`, confirm dialog is displayed and execution result
is saved in log file automatically. This log file double with lock file for mutual exclusion
and preventing rerunning.

=head1 CAUTION

THE SOFTWARE IS IT'S IN ALPHA QUALITY. IT MAY CHANGE THE API WITHOUT NOTICE.

If forking in your script, the software may not be properly handling it.

=head1 INTERFACE

=head2 Functions

=head3 C<< hissatsu(%options) >>

Automatically called in use phase.

Available options are:

=head4 C<< exec_date => 'Str|Time::Piece' >>

Date for execution date. Format is '%Y-%m-%d'.

=head4 C<< confirm_dialog => 'Bool' >>

Confirm dialog is to be displayed or not.
default: 1

=head4 C<< dialog_message => 'Str' >>

Message of confirm dialog.
Script name is expanded to '%s'.
If using multibyte strings, you should C<$ use utf8;> before C<$ use Script::Ichigeki>.
default: 'Do you really execute `%s` ?',

=head1 DEPENDENCIES

Perl 5.8.1 or later.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please either email me, or add the bug
to cpan-RT.

=head1 SEE ALSO

L<perl>

=head1 AUTHOR

Masayuki Matsuki E<lt>y.songmu@gmail.comE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2012, Masayuki Matsuki. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
